rust/kernel/sync/arc.rs

```diff
@@ -7,178 +7,216 @@
-    pub unsafe fn from_raw(ptr: *const T) -> Self {
+    /// `encoded` must have been returned by a previous call to [`Ref::into_usize`]. Additionally,
+    /// it can only be called once for each previous call to [``Ref::into_usize`].
+    pub unsafe fn from_usize(encoded: usize) -> Self {
         Ref {
-            ptr: NonNull::new(ptr as _).unwrap(),
+            ptr: NonNull::new(encoded as _).unwrap(),
         }
     }
 }
 
-impl<T: RefCounted + ?Sized> Deref for Ref<T> {
+impl<T: ?Sized> Ref<T> {
+    /// Determines if two reference-counted pointers point to the same underlying instance of `T`.
+    pub fn ptr_eq(a: &Self, b: &Self) -> bool {
+        core::ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
+    }
+
+    /// Returns a pinned version of a given `Ref` instance.
+    pub fn pinned(obj: Self) -> Pin<Self> {
+        // SAFETY: The type invariants guarantee that the value is pinned.
+        unsafe { Pin::new_unchecked(obj) }
+    }
+}
+
+impl<T: ?Sized> Deref for Ref<T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
         // SAFETY: By the type invariant, there is necessarily a reference to the object, so it is
         // safe to dereference it.
-        unsafe { self.ptr.as_ref() }
+        unsafe { &self.ptr.as_ref().data }
     }
 }
 
-impl<T: RefCounted + ?Sized> Clone for Ref<T> {
+impl<T: ?Sized> Clone for Ref<T> {
     fn clone(&self) -> Self {
-        Self::new_from(self)
+        // INVARIANT: C `refcount_inc` saturates the refcount, so it cannot overflow to zero.
+        // SAFETY: By the type invariant, there is necessarily a reference to the object, so it is
+        // safe to increment the refcount.
+        unsafe { rust_helper_refcount_inc(self.ptr.as_ref().refcount.get()) };
+        Self { ptr: self.ptr }
     }
 }
 
-impl<T: RefCounted + ?Sized> Drop for Ref<T> {
+impl<T: ?Sized> AsRef<T> for Ref<T> {
+    fn as_ref(&self) -> &T {
+        // SAFETY: By the type invariant, there is necessarily a reference to the object, so it is
+        // safe to dereference it.
+        unsafe { &self.ptr.as_ref().data }
+    }
+}
+
+impl<T: ?Sized> Drop for Ref<T> {
     fn drop(&mut self) {
         {
             // SAFETY: By the type invariant, there is necessarily a reference to the object.
-            let obj = unsafe { self.ptr.as_ref() };
-
-            // Synchronises with the acquire below or with the acquire in `get_mut`.
-            if obj.get_count().count.fetch_sub(1, Ordering::Release) != 1 {
+            let inner = unsafe { self.ptr.as_ref() };
+
+            // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
+            // this instance is being dropped, so the broken invariant is not observable.
+            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount. We
+            // also need to limit the scope of `obj` as it cannot be accessed after the refcount is
+            // decremented.
```

Comment:
I need to clarify that this applies only when the count is non-zero. Let's say the count went to n (> 0) after the decrement; I cannot dereference the pointer anymore because other threads/CPUs may concurrently drop it to zero and free it.

If the count reaches zero, then no other thread/CPU can possibly be using it, and the current thread/CPU is tasked with freeing it.

I'll update the comment.

Reply:
As I was updating the code, I noticed that the comment is actually correct as I'm talking about the scope of the local reference.

I need all instances of `self.ptr.as_ref()` to be dropped before calling `Box::from_raw(self.ptr.as_ptr())` otherwise I'll have a `T` and `&T` at the same time (for the same object), which is a violation of aliasing rules.

Reply:
Technically, there's nothing wrong with keeping the `inner` reference alive for the whole duration of the function, as long as no-one "touches it" when it's invalid. But I agree that it's better to censor `inner` starting from the point where it becomes unsound to touch it.

I am guessing that in the current code, `inner` lives slightly too long. Would it be beneficial to drop it right after it becomes unsound to touch? For example:
```rust
    fn drop(&mut self) {
        let is_zero = {
            // SAFETY: By the type invariant, there is necessarily a reference to the object.
            let inner = unsafe { self.ptr.as_ref() };
            // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
            // this instance is being dropped, so the broken invariant is not observable.
            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
            // `inner` must no longer be accessed after the refcount is decremented.
            unsafe { rust_helper_refcount_dec_and_test(inner.refcount.get()) }
        };
        if !is_zero {
            return;
        }

        // The count reached zero, we must free the memory.
        //
        // SAFETY: The pointer was initialised from the result of `Box::leak`.
        unsafe { Box::from_raw(self.ptr.as_ptr()) };
    }
```

Reply:
> Technically, there's nothing wrong with keeping the `inner` reference alive for the whole duration of the function, as long as no-one "touches it" when it's invalid. But I agree that it's better to censor `inner` starting from the point where it becomes unsound to touch it.

Well, the argument about not touching may be true for C, not really for Rust. The mere _existence_ of incompatible references is considered UB, even if one of them is never touched. For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=017b2973a26351f92213c5d981978919)):

```rust
fn main() {
    let v = 3u32;
    let v_mut = unsafe { &mut *((&v as *const u32) as *mut u32) }; 
    println!("{}", v)
}
```

We never touch `v_mut`, but Miri flags this as UB:

```
error: Undefined Behavior: trying to reborrow for Unique at alloc1426, but parent tag <untagged> does not have an appropriate item in the borrow stack
 --> src/main.rs:3:26
  |
3 |     let v_mut = unsafe { &mut *((&v as *const u32) as *mut u32) }; 
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for Unique at alloc1426, but parent tag <untagged> does not have an appropriate item in the borrow stack
  |
```


> 
> I am guessing that in the current code, `inner` lives slightly too long. Would it be beneficial to drop it right after it becomes unsound to touch? For example:
> 
> ```rust
>     fn drop(&mut self) {
>         let is_zero = {
>             // SAFETY: By the type invariant, there is necessarily a reference to the object.
>             let inner = unsafe { self.ptr.as_ref() };
>             // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
>             // this instance is being dropped, so the broken invariant is not observable.
>             // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
>             // `inner` must no longer be accessed after the refcount is decremented.
>             unsafe { rust_helper_refcount_dec_and_test(inner.refcount.get()) }
>         };
>         if !is_zero {
>             return;
>         }
> 
>         // The count reached zero, we must free the memory.
>         //
>         // SAFETY: The pointer was initialised from the result of `Box::leak`.
>         unsafe { Box::from_raw(self.ptr.as_ptr()) };
>     }
> ```

You're right, it lives for too long. In fact, even your suggested version also has UB: after `rust_helper_refcount_dec_and_test` we still have `inner` (`&InnerRef<T>`) while another thread may have freed it. So I think we need to avoid it altogether while calling `rust_helper_refcount_dec_and_test`. Something like:

```rust
    fn drop(&mut self) {
        // SAFETY: By the type invariant, there is necessarily a reference to the object.
        let refcount = unsafe { self.ptr.as_ref() }.refcount.get();

        // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
        // this instance is being dropped, so the broken invariant is not observable.
        // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(refcount) };
        if is_zero {
            // The count reached zero, we must free the memory.
            //
            // SAFETY: The pointer was initialised from the result of `Box::leak`.
            unsafe { Box::from_raw(self.ptr.as_ptr()) };
        }
    }
```

Since `refcount` is a raw pointer, it's not violating any aliasing rules.

Reply:
That's really fascinating, I had never thought about aliasing rules. Given that I'm writing unsafe Rust now, I better get up to speed quickly :)

I'm not sure if we are invalidating any aliasing rules simply by holding incompatible references (i.e. never dereffing them). For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fa6a11c6caa11d25a68943113cc16f84)):
```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let _shared_ref = unsafe { ptr.as_ref() };
    unsafe { Box::from_raw(ptr.as_ptr()); }
    println!("I'm not touching _shared_ref");
    //println!("{}", _shared_ref);
}
```
Miri seems ok with the simple existence of the invalid `_shared_ref`, but it "has a cow" once you deref it.

That does sound compatible with the Rust aliasing rules as explained in the [`UnsafeCell` docs](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html): (emphasis mine)
> If you create a safe reference with lifetime 'a (either a &T or &mut T reference) that is accessible by safe code (for example, because you returned it), **then you must not access the data** in any way that contradicts that reference for the remainder of 'a. For example, this means that if you take the *mut T from an UnsafeCell<T> and cast it to an &T, then the data in T must remain immutable (modulo any UnsafeCell data found within T, of course) until that reference’s lifetime expires. Similarly, if you create a &mut T reference that is released to safe code, then you must not access the data within the UnsafeCell until that reference expires.

Is the goal here to prevent invalid references from ever existing? If not, could we simplify things here by not trying to limit the reference's lifetime?

But if so, then the current code is still not 100% in compliance: the `refcount: *mut bindings::refcount_t` pointer is not always valid:
```rust
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(refcount) };
        // refcount invalid right here
        if is_zero {
            // refcount valid here
```
Perhaps the only way to solve this problem is by not creating the reference in the first place?
```rust
    fn drop(&mut self) {
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(self.ptr.as_ref().refcount.get()) };
        if is_zero {
            unsafe { Box::from_raw(self.ptr.as_ptr()) };
        }
    }
```

Reply:
> I'm not sure if we are violating any aliasing rules simply by holding invalid references (i.e. never dereffing them). For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fa6a11c6caa11d25a68943113cc16f84)):
> 
> ```rust
> use std::ptr::NonNull;
> use std::boxed::Box;
> 
> fn main() {
>     let ptr = NonNull::from(Box::leak(Box::new(3u32)));
>     let _shared_ref = unsafe { ptr.as_ref() };
>     unsafe { Box::from_raw(ptr.as_ptr()); }
>     println!("I'm not touching _shared_ref");
>     //println!("{}", _shared_ref);
> }
> ```
> 
> Miri seems ok with the simple existence of the invalid `_shared_ref`, but it "has a cow" once you deref it.

It looks like the compiler is taking advantage of [non-lexical lifetimes](https://doc.rust-lang.org/edition-guide/rust-2018/ownership-and-lifetimes/non-lexical-lifetimes.html) to forget `_shared_ref` early. See this example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d35eb74a0de53974b00d150da6e12590)):

```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn f(ptr: NonNull<u32>, b: &u32) {
    unsafe { Box::from_raw(ptr.as_ptr()); }
}

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let shared_ref = unsafe { ptr.as_ref() };
    f(ptr, shared_ref);
}
```

By your standards, `shared_ref` is not touched either, but Miri doesn't like it.

> Is the goal here to prevent invalid references from ever existing? If not, could we simplify things here by not trying to limit the reference's lifetime?

The goal is to prevent conflicting/dangling/invalid references from ever existing.

> But if so, then the current code is still not 100% in compliance: the `refcount: *mut bindings::refcount_t` pointer is not always valid:

It's ok to have a dangling/invalid raw pointers (`*const T` and `*mut T`) as long as you don't dereference them. It's not ok to have a dangling/invalid references (`&T` and `&mut T`).

See for example [`addr_of!`](https://doc.rust-lang.org/beta/std/ptr/macro.addr_of.html), they defined it to avoid creating an invalid reference while going from a raw pointer to another raw pointer:

> `&expr as *const _` creates a reference before casting it to a raw pointer, and that reference is subject to the same rules as all other references. This macro can create a raw pointer _without_ creating a reference first.

> Perhaps the only way to solve this problem is by dropping the reference before `refcount_dec_and_test()` returns?

I think this is fine too, but we have two unsafe operations in a single block. I want to separate them out.

Reply:
I'm not sure what to make of the Rust aliasing rules as described in the `UnsafeCell` docs  then. I'll make abstraction of that for now :)

So yes, let's use the code we have right now, which keeps a `refcount: *mut bindings::refcount_t` around, which is invalid at certain points in the function.

Would you be able to humour me, and document briefly that a) we cannot touch the pointer after the call, and b) it's not a violation of Rust's aliasing rules?

Reply:
> By your standards, shared_ref is not touched either, but Miri doesn't like it.

A local reference is different from a reference in the function parameter. A function parameter is "protected" (in Miri terminology) and is marked as `dereferenceable` for LLVM, so LLVM is free to dereference it anywhere inside the function. On the other hand, with NLL, a local reference cannot be freely dereferenced by LLVM after the last usage.

Reply:
> A local reference is different from a reference in the function parameter. A function parameter is "protected" (in Miri terminology) and is marked as `dereferenceable` for LLVM, so LLVM is free to dereference it anywhere inside the function. On the other hand, with NLL, a local reference cannot be freely dereferenced by LLVM after the last usage.

Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

Reply:
> I'm not sure what to make of the Rust aliasing rules as described in the `UnsafeCell` docs then. I'll make abstraction of that for now :)

What part of that do you think is contradictory to what I'm saying? My reading of it is: if you release a reference to safe code, then you (who hold a raw pointer) must not touch the data in incompatible ways (because safe code  may be doing it).

Further down they have the following:

> To assist with proper design, the following scenarios are explicitly declared legal for single-threaded code:
> 1. A `&T` reference can be released to safe code and there it can co-exist with other `&T` references, but not with a `&mut T`
> 1. A `&mut T` reference may be released to safe code provided neither other `&mut T` nor `&T` co-exist with it. A `&mut T` must always be unique.

Note the use of _co-exist_. They don't say anything about dereferencing.

> So yes, let's use the code we have right now, which keeps a `refcount: *mut bindings::refcount_t` around, which is invalid at certain points in the function.
> 
> Would you be able to humour me, and document briefly that a) we cannot touch the pointer after the call, and b) it's not a violation of Rust's aliasing rules?

Of course, I'll upload it shortly.

Reply:
> Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

No. Here's another example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=945914d08b0c1c2793e6921913347b6f)) without a function:

```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let shared_ref = unsafe { ptr.as_ref() };
    unsafe { Box::from_raw(ptr.as_ptr()); }
    println!("I'm not touching _shared_ref");
    let another_ref = shared_ref;
}
```

Here neither `shared_ref` nor `another_ref` are ever dereferenced. But the usage of `shared_ref` afterwards prevents NLL from limiting `shared_ref`'s lifetime. And miri flags UB.

Reply:
> Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

It's possible for invalid reference to exist even in safe Rust:
```rust
let mut v = vec![1];
let x = &v[0];
v.push(2);
// now x is invalid reference
```
This is accepted by NLL. Obviously borrow checking will prevent you from accessing `x` ever again, but after borrow checking all lifetimes are erased, and for codegen `x` is just an invalid reference after push.

> ```rust
> let another_ref = shared_ref;
> ```

This is considered as an use. If you remove this then it's not UB.

Reply:
So we have a consensus now? We'll keep a `refcount: *mut bindings::refcount_t` in scope, although it's not always valid, which nevertheless does not violate Rust's aliasing rules. And we briefly document all of this in the code.

Except @nbdd0121 if you'd like to suggest something better?

Reply:
> > Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?
> 
> It's possible for invalid reference to exist even in safe Rust:
> 
> ```rust
> let mut v = vec![1];
> let x = &v[0];
> v.push(2);
> // now x is invalid reference
> ```
> 
> This is accepted by NLL. Obviously borrow checking will prevent you from accessing `x` ever again, but after borrow checking all lifetimes are erased, and for codegen `x` is just an invalid reference after push.

Right, if something is inaccessible it obviously doesn't count for the aliasing rules. If it did, something like:

```rust
fn f(p: &u32) {
    todo!()
}

fn g(q: &mut u32) {
    f(q)
}
```
Would be a violation of aliasing rules because you have a `&T` and `&mut T` at the same time. It isn't because `q: &mut u32` becomes inaccessible during the call to `f`.
 
> 
> > ```rust
> > let another_ref = shared_ref;
> > ```
> 
> This is considered as an use. If you remove this then it's not UB.

It's an use, of course, but not a dereferencing of the pointer.

Reply:
> So we have a consensus now? We'll keep a `refcount: *mut bindings::refcount_t` in scope, although it's not always valid, which nevertheless does not violate Rust's aliasing rules. And we briefly document all of this in the code.

The current approach looks good to me. I just want to point out that it is *okay* to have a local (but not as function parameter) dangling/invalid reference as long as it is not used, but obviously without borrow checker doing the enforcement it is *preferable* to avoid having it in scope at all.